packages/backend-common/src/reading/types.ts

```diff
@@ -22,6 +22,7 @@ import { Config } from '@backstage/config';
  */
 export type UrlReader = {
   read(url: string): Promise<Buffer>;
+  readTree(url: string): Promise<string>;
```

Comment:
Not sure about what we actually want this signature to be, I'm thinking there are a few signatures that could make sense in isolation, but a bit conflicted if we should pick one or combine.

First option would be a in-memory representation of some sort, which is what you implemented. Details of the `File` type don't matter too much at this point:

```ts
type File = {
  path: string;
  content: Buffer;
  mode: string
}

readTree(url: string): Promise<File[]>
```

Another option is returning a `.tar.gz` file. I think this can be a handy one in plenty of cases. You'll end up with a smaller memory footprint and there's plenty of utilities around extracting parts of the archive, it's easy to send to other services, and so on. It's is a bit awkward that you always have to bring in external libs to work with the response though.

```ts
readTree(url: string): Promise<{archive: Buffer}> // a .tar.gz archive
```

Third and final option that I can come up with is placing the files in a directory. This is another one that's even closer to the actual usage as it's essentially what both techdocs and the scaffolder does. It may be a bit slow to always reach for filesystem operations though.

```ts
// outDir could be used to place the contents in a specific dir, and when omitted tmpdir will be used
readTree(url: string, outDir?: string): Promise<{dir: string}>
```

I think there's benefits and drawbacks to all of these, and there might be other options too. None of them stand out as an obvious choice tbh.

Thinking about if we could combine them I realized that modelling it similar to `fetch` responses could be a sane option. Something like this:

```ts
type File = {
  path: string
  content(): Promise<Buffer>
}

type ReadTreeResponse = {
  files(): Promise<File[]>
  archive(): Promise<Buffer>
  dir(outDir?: string): Promise<string>
}

readTree(url: string): Promise<ReadTreeResponse>;
```

That'd let us decouple the implementation from the output method too, avoiding possible things like fetching dir as a tarball, then extracting that to a directory, just to have the caller bundle it all up into a tarball again. It could also let us expand with additional metadata in the future, such as `commitId(): string | undefined` or `etag(): string` 

@hooloovooo @benjdlambert @freben thoughts on ^?

Btw, we should totally also redo the response type of `read` if we go with that, just to allow it to evolve more in the future.

Reply:
I think the last example sounds like a good idea. Would allow more flexibility and as you said adding things like commit hashes which would be very useful for techdocs.

Reply:
Yep me too. I'm just catching up on all this `readTree` stuff but it seems sensible to do it that way.